### ISSUE TITLE
WithStatus instances are _still_ overlapping (sorry :( )

### ISSUE THREAD

**Issue by arianvp** [haskell-servant/servant#1367]
From a collegue on `0.18.2`

```
 • Overlapping instances for MimeRender JSON (WithStatus 201 ())
  arising from a use of ‘Servant.serve’
Matching instances:
  two instances involving out-of-scope types
    instance MimeRender ctype a => MimeRender ctype (WithStatus _status a)
    instance [overlappable] ToJSON a => MimeRender JSON a
```

Even without the `ToJSON (WithStatus a)`  instance GHC is still convinced these two overlap.

I am confused with it would think there is an overlap though. Note that there is _no_ `ToJSON (WithStatus 201 ())` instance at all so why it even considers the `  instance [overlappable] ToJSON a => MimeRender JSON a` instance is baffling me.  Maybe the other instance needs an `{-# OVERLAPPING #-}` pragma? Why though?

This is trivially caught if the `uverb` cookbook would call the `serve` function (which it currently doesn't). And I didnt compile the cookbook locally either (as the `cabal.project` file for the cookbooks differs from the one travis generates; and it isn't working).

Action points:

* Make the cookbook fail to compile by calling `serve` on the API type. So that we do not ship these things without proper testing anymore
* Make the cookbooks compile locally again (They currently do not as the `cabal.project` for the cookbooks is self-inconsisent.  CI doesn't mind as it generates its own `cabal.project`)
* Figure out why it overlaps

**Comment by arianvp** [comment]
Apparently instance context is not considered for selection.

```
MimeRender ctype (WithStatus _status a)
MimeRender JSON a
```

and neither are more specific

**Comment by Taneb** [comment]
The downside of that approach is that it necessitates modification of every library that provides a content type for servant to work with `WithStatus`

**Comment by arianvp** [comment]
Indeed a bit of a bummer; but I am afraid there is no way around it.

Note that the `WithStatus` newtype is just there as a convenience; you do not need it to use UVerb.
We should stress this more in the documentation.

You can give your own datatype an instance of `HasStatus`:
```haskell
data MyType = MyType { ... } deriving HasStatus via (WithStatus 300 MyType)

-- or  without derivingvia:
instance HasStatus  MyType where type Status MyType = 300

type MyAPI = UVerb GET  '[MyType, MyOtherType]
```

**Comment by intolerable** [comment]
I don't think there's a way of doing it than defining the `MimeRender` instance for `WithStatus` for each content type, unfortunately (a nice one, anyway).

Is it worth adding a new test in `ServerSpec` that tries to serve an API that contains a `WithStatus` with content type JSON?

**Comment by intolerable** [comment]
I just saw your comment in the other thread, I'll add the test to a PR on this issue

**Comment by fisx** [comment]
(also tested against https://github.com/haskell-servant/servant/pull/1368)

**Comment by gdeest** [comment]
Do I understand correctly that this issue was accidentally closed by #1376 , but doesn't actually solve the issue ? 

I managed to reproduce a similar problem on top of #1388 , by defining routes as _polymorphic_ records.
